sdk/core/core/inc/az_const_str.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_CONST_STR_H
+#define AZ_CONST_STR_H
+
+#include <az_assert.h>
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// A constant string.
+typedef struct {
+  // Points to the first character.
+  char const *begin;
+  // A numer of C characters (bytes) in the string.
+  size_t size;
+} az_const_str;
+
+typedef struct {
+  char* begin;
+  size_t size;
+} az_str;
+
+// A size of the string literal.
+// Details: to make sure that `S` is a `string literal`, we are appending `""` to `S`.
+#define AZ_STRING_LITERAL_SIZE(S) (sizeof(S "") - 1)
```

Comment:
this is asking for bugs when people use this size as a parameter to memmove. Maybe name it "LENGTH" or at least put in the comment that it does not include the null.